src/libstore/machines.cc

```diff
@@ -6,29 +6,24 @@
 
 namespace nix {
 
-Machine::Machine(decltype(storeUri) storeUri,
+Machine::Machine(
+    const std::string & storeUri,
     decltype(systemTypes) systemTypes,
     decltype(sshKey) sshKey,
     decltype(maxJobs) maxJobs,
     decltype(speedFactor) speedFactor,
     decltype(supportedFeatures) supportedFeatures,
     decltype(mandatoryFeatures) mandatoryFeatures,
     decltype(sshPublicHostKey) sshPublicHostKey) :
-    storeUri(
-        // Backwards compatibility: if the URI is schemeless, is not a path,
-        // and is not one of the special store connection words, prepend
-        // ssh://.
-        storeUri.find("://") != std::string::npos
-        || storeUri.find("/") != std::string::npos
-        || storeUri == "auto"
-        || storeUri == "daemon"
-        || storeUri == "local"
-        || hasPrefix(storeUri, "auto?")
-        || hasPrefix(storeUri, "daemon?")
-        || hasPrefix(storeUri, "local?")
-        || hasPrefix(storeUri, "?")
-        ? storeUri
-        : "ssh://" + storeUri),
+    storeUri([&storeUri]{
+        try {
+            return StoreReference::parse(storeUri);
+        } catch (UsageError &) {
```

Comment:
This is nasty. Should we deprecate this old syntax?

Reply:
Could this be done during config parsing instead? That way `nix show-config` will show the correct syntax.

Also no chance to accidentally keep using this hack when we improve the settings format.

Reply:
I would like to parse more eagerly, but that felt like a change best kept for later. (Similarly, using `Setting<StoreReference>` for a bunch of things.)

Right now, this is just supposed to match the back compat hack that was there before.

Reply:
I just restored the old method of compat for now. We can decide separately whether such try-catch is worth it.